### DISCUSSION TITLE
Hybrid search usin vector indx and inverted index

### DISCUSSION THREAD

**Discussion by patelprateek** [milvus-io/milvus#36940] [Q&A and General discussion]
Could some one help me understand how hybrid search works in milvus and point me to some code pointers.
For ex : i want to get topk similar items constrained by some where clause. Do we do pre-filtering or post filtering ?
I am trying to use hnsw , which stores its own internal ids for documents , do we translate inverted index ids to hnsw ids (bitset) when doing pre-filtering , does this method scale to billion scale index.

**Comment by yhmo** [comment]
https://milvus.io/docs/reranking.md#Weighted-Scoring-WeightedRanker
The doc mentioned "For instance, the distance for IP ranges from [-∞,+∞], while the distance for L2 ranges from [0,+∞]. Milvus employs the arctan function, transforming values to the [0,1] range to provide a standardized basis for different metric types.".
Arctan is for IP and L2 whose range is unlimited(∞).
COSINE range is [-1, 1], is limited, the (1 + distance) * 0.5 makes sense.
- If metric type is Cosine: distance = (1 + distance) * 0.5
- If metric type is IP: distance = 0.5 + float32(math.Atan(float64(distance)))/math.Pi
- If metric type is L2: distance = 1.0 - 2*float32(math.Atan(float64(distance)))/math.Pi

Let's say we have two vector fields(A and B) in a collection, now we conduct a hybrid search with WeightedRanker(0.2, 0.8).
The metric type is COSINE. From A we get top3 items, from B we get top3 items, and their distance.
```
Top3 of A: No.1: {id=8, distance=0.9}   No.2: {id=5, distance=0.7}   No.3: {id=1, distance=0.5}
Top3 of B: No.1: {id=7, distance=0.8}   No.2: {id=1, distance=0.6}   No.3: {id=8, distance=0.3}
```

1. First step: transfer the distance to range [0, 1] by distance = (1 + distance) * 0.5
```
Top3 of A: No.1: {id=8, distance=0.95}   No.2: {id=5, distance=0.85}   No.3: {id=1, distance=0.75}
Top3 of B: No.1: {id=7, distance=0.9}   No.2: {id=1, distance=0.8}   No.3: {id=8, distance=0.65}
```

2. Second step: use weights to calculate each id's hybrid-distance:
```
id=8: 0.95 * 0.2 + 0.65 * 0.8 = 0.71
id=5: 0.85 * 0.2 + 0.0 * 0.8 =  0.17  // id=5 doesn't in the top3 of B, so the second distance is 0.0
id=1: 0.75 * 0.2 + 0.8 * 0.8 = 0.79
id=7: 0.0 * 0.2 + 0.9 * 0.8 = 0.72    // id=7 doesn't in the top3 of A, so the first distance is 0.0
```

3. Third step: rerank to get the final top3 items
```
No.1: {id=1, distance=0.79}
No.2: {id=7, distance=0.72}
No.3: {id=8, distance=0.71}
```

Here is a discussion about the internal algorithm of hybrid search:
https://github.com/milvus-io/milvus/discussions/34415

**Comment by yhmo** [comment]
Milvus is pre-filtering. Before ANN search, the filtering expression is parsed and executed to return a bitset. In this bitset, each bit represent an item, 1 is to compute, 0 is to skip. https://milvus.io/docs/bitset.md#Bitset

Milvus manages data by segments, each segment has an independent index. When it receives a search request, the request is applied to each segment. Filtering expression is applied first to get a bitset of the segment, then the search engine only computes the distance of the items whose bit is 1.

Since data is split into segments, it is easy to distribute a large number of segments to lots of query nodes, and the service is easy to scalable.

**Comment by patelprateek** [comment]
@yhmo : thanks for the response . Couple of followups
1) My understanding based on your response is we partition data into segment and for data in a segment we build multiple index i.e ann index and inverted index ( and may be more). is this correct ? if true then how big are typical segments ? It seems to me this is inefficient for ann inddex where its better to have as many items in segment  ?
2) Since hnsw has its own internal id , do we pre-filter and generate bitset , but this bitset needs to be based on internal id mapping that hnsw keeps ? Or both inverted index and ann index need to have exact same id mapping else the bitset will not be useful ? does this method scale to approach where my pre-filtering result may generate millions of documents that then need to be translated to bitsets and then passed to ann search ?
3) how do we support incremental or streaming ingestion and indexing  ?

**Reply by yhmo** [reply]
The size of a segment is configured in the milvus.yaml:
For normal indexes: https://milvus.io/docs/configure_datacoord.md#dataCoordsegmentmaxSize, default value is 1024MB
For DiskANN index: https://milvus.io/docs/configure_datacoord.md#dataCoordsegmentdiskSegmentMaxSize, default value is 2048MB

In runtime, some segments might be larger than maxSize, some others might be smaller than maxSize. Averagely the size of per segment is near to maxSize.

For each segment, there must be at least a vector field. For each vector field, there is an independent index. In current version, one vector field only has one index type.

All the pre-filter/mapping work is already covered by the search engine, not visible to users. The search engine knows which internal id is true or false in the bitset. Users only know the primary key of the collection. For each collection, there must be a primary key field,  which means each item has a unique primary ID, the search engine knows which primary ID is mapped to which internal hnsw id.

Milvus is a complicated system, I could not tell all the details, I just describe a brief workflow of milvus:

Milvus manages data by segments. By default, each segment's size is expected to be 1024MB(in v2.3, the default size is 512MB).

A segment is persisted as multiple chunks. In each chunk, a column-based file for each field.
One collection can contain multiple partitions, one partitions can contain multiple segments. One segment contains multiple chunks. One chunk contains multiple column-base data files.
By default, each collection has a default partition.

There are two kinds of segments:
- Growing segment, can accept new data, some chunks are persisted into S3/MinIO, the last chunk is in-memory buffer to accept new data.
- Sealed segment, data is static, cannot be changed, all chunks are persisted into S3/MinIO. Small sealed segments can be merged into large segments.

A milvus cluster has these components: https://milvus.io/docs/main_components.md

Responsibility for each component:
- Pulsar/Kafka acts as a message queue to store WAL(write-ahead-log) for insert/delete requests, and time-tick messages for data consistency. All the inserted data will be stored in Pulsar/Kafka before being consumed by datanode/querynode.
- MinIO/S3 is the major data storage of Milvus, all the original data and index data is stored here. The data is organized as collection/partition/segment/field. Data is stored as column-based.
- Etcd stores the system metadata of Milvus, including schema of collections, runtime address of each component, paths of data/index of each segment, etc.
- Proxy receives requests from clients and dispatches requests to other components.
- Query node loads segments into memory for search, including growing segments and sealed segments. 
- Data node persists data into segments, merges small segments into large segments.
- Index node builds indexes for sealed segments.
- Root coordinator maintains data consistency and the entire system state. Milvus manages data consistency by time-tick machinery: https://milvus.io/docs/time_sync.md#Time-Synchronization. The root coordinator sends a time-tick for each 200ms, and each "data-channel" will receive the time-tick messages. Once a time-tick message is accepted by all the datanodes and querynodes, we can say the data before this time point is visible and searchable.
- Data coordinator manages data nodes, and assigns tasks to data nodes.
- Query coordinator manages query nodes and index nodes(in old versions, index coordinator manages index nodes, in new version, index coordinator is removed)


For each partition, at least one growing segment is created. When you create a collection, there is a parameter shards_num, the default value is 1, which means each partition has one "data-channel" to accept new data. A "data-channel" is a virtual data channel to consume data from Pulsar/Kafka to datanode or querynode.
In milvus, if a collection is loaded, all the segments will be loaded into query nodes, including growing segments and sealed segments.


When you insert some data by insert():
1. The data is received by proxy node, proxy node splits the data into pieces by shards_num(which is defined in collection schema, default is 1), and sends the pieces(as messages) into Pulsar/Kafka. The insert() returns success to the client once the data is persisted into Pulsar/Kafka.
2. For each growing segment, two "data channels" are created, one is between querynode and Pulsar/Kafka, the other is between datanode and Pulsar/Kafka.
3. Querynode and datanode subscribe messages asynchronously from Pulsar/Kafka, the data pieces are consumed by querynode and datanode.
4. For each growing segment, datanode keeps a buffer to accept new data. Once the buffer size reaches 16MB or the buffer lifetime hits the limit(10 minutes), the buffer is persisted to be a chunk into S3/MinIO.
5. If the collection is loaded, all segments are kept in query nodes. for each growing segment, querynode keeps a buffer to accept new data, so that search requests can search the data. New data is consumed by querynode from Pulsar/Kafka, only when the new data is already consumed by querynode, the data is visible&searchable by search()/query(). 
6. With more and more chunks persisted, the size of the growing segment reaches a threshold(100+MB), the growing segment is converted to be sealed segment by data coordinator. No more data is allowed to be inserted into this segment. Query node and data node will create a new buffer for the next growing segment.
7. Once a new sealed segment is generated, index node will receive a task to build index for this seal segment.
8. Once an index is finished, the query node will load this index into memory to replace the original segment data.


When you call load(), there is a parameter "replica_number", default value is 1. This parameter determines how many copies of this collection are loaded in memory. If you set replica_number=2, there will be 2 copies in memory, and Milvus can balance search requests between the 2 copies. For each copy, there is a query node to lead/manage requests.


When you call search() on the client side:
1. A search request is sent to a proxy node of Milvus
2. The proxy node passes the request to a "delegator query node" which is leading a copy, the query node knows the distribution of all the segments of this copy. Some segments might be loaded in other query nodes. The "delegator" passes the request to other query nodes, gets the results from other query nodes, combines the results to a final result and returns the the proxy.
3. The proxy returns the final result to the client

**Reply by yhmo** [reply]
- "how do we support incremental or streaming ingestion and indexing ?"

Based on the milvus machinery, it is easy for users to insert data, no matter streaming insert or batch insert. The milvus server can merge the streaming data into segments, and automatically build index for each segment.

For example, both streaming insert or batch insert are ok:
```
# streaming insert
for i in range(1000000):
    collection.insert({"id": 1, "vector": [......]})   # each insert request inserts one row

# batch insert
data = [
  [1, 2, 3, ..... 1000000],  # id list
  [vector1, vector2, .... vector100000],  # vector list
]
collection.insert(data)    # one insert request with 1000000 rows
```

Users just following this workflow of milvus to insert/load/search:
```
create_collection()
collection.insert()
collection.create_index()
collection.load()
collection.search()

```
